Removal of s.el as a dependency

dotemacs opened the pull request:
This PR based on the this issue https://github.com/clojure-emacs/clj-refactor.el/issues/293 "Removal of s.el as a dependency".

The dependency on s.el has now been removed.

Please provide your feedback about the work I've done so far.

---- 

**Regarding the PR etiquette below, I've crossed out the sections that I don't think are applicable.**

Before submitting a PR make sure the following things have been done (and denote this
by checking the relevant checkboxes):

- [x] The commits are consistent with our [contribution guidelines](https://github.com/clojure-emacs/clj-refactor.el/blob/master/.github/CONTRIBUTING.md)
- [x] You've added tests (if possible) to cover your change(s)
- [x] The new code is not generating byte compile warnings (run `cask exec emacs -batch -Q -L . -eval "(progn (setq byte-compile-error-on-warn t) (batch-byte-compile))" clj-refactor.el`)
- [x] All tests are passing (run `./run-tests.sh`)
~~- [ ] You've updated the changelog (if adding/changing user-visible functionality)~~
~~- [ ] You've updated the readme (if adding/changing user-visible functionality)~~

Thanks!

dotemacs commented on clj-refactor.el:
```diff
@@ -3365,6 +3366,17 @@ See: https://github.com/clojure-emacs/clj-refactor.el/wiki/cljr-create-fn-from-e
+    (while (not (eobp))
+      (subword-forward)
+      (insert " "))
+    (mapconcat 'identity (split-string (downcase (buffer-string))) "-")))
```
What do you think of this?

I'm a bit undecided. 

In one way it only uses functions that come as part of Emacs and relies on `subword` mode. But a bit unsure about it... suggestions welcome.

expez replied:
Looks OK to me.  If it works as it should I don't see a huge point in bikeshedding implementation details we can change at will :)

dotemacs replied:
> Looks OK to me.

Cool, glad that you approve :)

dotemacs commented:
The only functions left so far from **s.el** are `s-slice-at` & `s-with`. 
They are both used only once.

I'll try to tackle those tonight/tomorrow...

dotemacs commented:
Only `s-slice-at` is left...

dotemacs commented:
I've removed all the functions from `s.el` and used stock ones from Emacs.

I've updated the PR description with some comments, please read that first as there are a few outstanding issues that I require your feedback on.

Then please review the code.

Thank you :)

benedekfazekas commented:
to  address your comments above in bold:

contributing guidelines here: https://github.com/clojure-emacs/clj-refactor.el/blob/master/.github/CONTRIBUTING.md perhaps the link is wrong in the PR template?!

the README also has the (correct) link

tests: this is internal refactoring i guess we need to rely on our existing test set to catch regressions... don't think you need to add new unit like tests for this.

benedekfazekas commented on clj-refactor.el:
```diff
@@ -573,7 +573,7 @@ list of (fn args) to pass to `apply''"
 
 (defun cljr--whitespacep (s)
   "True if S contains only whitespace."
-  (s-blank? (string-trim s)))
```
essential difference between `s-blank?` and `string-blank-p` that former is null safe latter is not. I don't think that causes any problem in this commit but you might want to dblcheck that..

dotemacs replied:
Good point, added the null safety with 36a886b

benedekfazekas commented on clj-refactor.el:
```diff
@@ -761,14 +761,17 @@ A new record is created to define this constructor."
+			    (locate-dominating-file default-directory "build.boot")))
+			 (ignore-errors (file-truename
+					 (locate-dominating-file default-directory "pom.xml"))))))
+    (if (null project-dir)
```
perhaps rather `(or project-dir "")`?

or in fact you can get rid of the `let` you introduced and just stick an empty string at the end of the original `or` expression as default?

dotemacs replied:
Good point.

Changed it as per your comment.

I was more on the 'drive' to complete the whole task, so I've missed the obvious stuff like this. Thanks

benedekfazekas commented on clj-refactor.el:
```diff
@@ -2689,9 +2689,9 @@ str/split => split"
     (cond
      ((string-match-p "\\w+\\.\\w+" name)
-      (thread-last name (s-split "\\.") last car cljr--symbol-suffix))
+      (thread-last name (split-string "\\.") last car cljr--symbol-suffix))
```
this should be `thread-first` as well, or?

dotemacs replied:
Good catch. Fixed.

Shame that the tests didn't catch this...

benedekfazekas replied:
tests: indeed

dotemacs replied:
I'd happily add the test for this, but seeing as `cljr--symbol-suffix` is a non-interactive function and most of the tests are feature tests, should I have a step like this:

```feature
When I press "M-: (cljr--symbol-suffix \"java.util.Date\")"
Then I should see message "Date"
```

What are your thoughts?

benedekfazekas replied:
we decided to go down the behavioural testing route not the unit testing one. which is essence was a good choice i think. so i suppose just leave it ;)

benedekfazekas commented on clj-refactor.el:
```diff
@@ -2247,15 +2247,15 @@ See: https://github.com/clojure-emacs/clj-refactor.el/wiki/cljr-update-project-d
             (mc/create-fake-cursor-at-point)))
         (re-search-forward "\\[")
-        (when (s-present? locals)
+	(when (not (or (null locals) (string= "" locals)))
```
i would possibly rather go for
`string-empty-p` instead of `string= ""` for readability and perhaps factor this pred out in its own function? something like `cljr--string-present-p`?

dotemacs replied:
Created `cljr--string-present-p` as per your comment...

benedekfazekas commented:
made couple of comments. hope they make sense. only one of those is a potential bug the rest is more like style etc things.

dotemacs commented:
@benedekfazekas addressed all your comments.

Please review again and let me know what you think.

Thanks :)

benedekfazekas commented on clj-refactor.el:
```diff
@@ -769,9 +769,7 @@ A new record is created to define this constructor."
-    (if (null project-dir)
-	""
-      project-dir)))
+    (or project-dir "")))
```
do you need the `let` at all tho? you can just
```
(or (ignore-errors...)
     (ignore-errors...)
     (ignore-errors...)
     "")
```

or am i missing something?

benedekfazekas commented:
made one more comment. otherwise LGTM. great work!

benedekfazekas commented:
one more thing: `cljr--string-present-p` could be used at more places i am guessing. for example line 2264 and 3207

dotemacs commented:
Reused `cljr--string-present-p` where suitable... 1641730
Removed let c02c15f & reduced repetition 4c0b23b

See what you think of it now. 

Thanks for the feedback

benedekfazekas commented on clj-refactor.el:
```diff
@@ -760,16 +760,18 @@ A new record is created to define this constructor."
+  (or
+   (thread-last  '("project.clj" "build.boot" "pom.xml")
+     (mapcar 'cljr--locate-project-file)
+     (delete '"")
```
really like this approach. but i think this line is wrong. `cljr--locate-project-file` returns nil not empty string if file is not present. should be `(delete nil)` instead.

dotemacs replied:
Thanks, pushed it as part of c33fc5f

benedekfazekas commented:
great work @dotemacs ta!
